src/Utopia/Messaging/Adapters/Push/APNS.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Utopia\Messaging\Adapters\Push;
+
+use Exception;
+use Utopia\Messaging\Adapters\Push as PushAdapter;
+use Utopia\Messaging\Messages\Push;
+
+class APNS extends PushAdapter
+{
+    /**
+     * @param  string  $authKey
+     * @param  string  $authKeyId
+     * @param  string  $teamId
+     * @param  string  $bundleId
+     * @param  string  $endpoint
+     * @return void
+     */
+    public function __construct(
+      private string $authKey,
+      private string $authKeyId,
+      private string $teamId,
+      private string $bundleId,
+      private string $endpoint
+    ) {
+    }
+
+    /**
+     * Get adapter name.
+     *
+     * @return string
+     */
+    public function getName(): string
+    {
+        return 'APNS';
+    }
+
+    /**
+     * Get max messages per request.
+     *
+     * @return int
+     */
+    public function getMaxMessagesPerRequest(): int
+    {
+        return 1000;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param  Push  $message
+     * @return string
+     *
+     * @throws Exception
+     */
+    public function process(Push $message): string
+    {
+        $payload = [
+            'aps' => [
+                'alert' => [
+                    'title' => $message->getTitle(),
+                    'body' => $message->getBody(),
+                ],
+                'badge' => $message->getBadge(),
+                'sound' => $message->getSound(),
+                'data' => $message->getData(),
+            ],
+        ];
+
+        return \json_encode($this->notify($message->getTo(), $payload));
+    }
+
+    private function notify(array $to, array $payload): array
+    {
+        $headers = [
+            'authorization: bearer '.$this->generateJwt(),
+            'apns-topic: '.$this->bundleId,
+        ];
+
+        $ch = curl_init();
+
+        curl_setopt($ch, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_0);
+
+        curl_setopt_array($ch, [
+            CURLOPT_PORT => 443,
+            CURLOPT_HTTPHEADER => $headers,
+            CURLOPT_POST => true,
+            CURLOPT_POSTFIELDS => \json_encode($payload),
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_TIMEOUT => 30,
+            CURLOPT_HEADER => true,
+        ]);
+
+        $response = '';
+
+        foreach ($to as $token) {
+            curl_setopt($ch, CURLOPT_URL, $this->endpoint.'/3/device/'.$token);
+
+            $response = curl_exec($ch);
+        }
```

Comment:
I don't think this will utilize multiplexing 🤔 we'll need to use [`curl_share_init()`](https://www.php.net/manual/en/function.curl-share-init.php) and a new `curl_init()` with the `CURLOPT_SHARE` option set for each request